cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapperWrapper.java

```diff
@@ -195,15 +204,18 @@ public InputSplit getInputSplit() {
     return flushingContext;
   }
 
-  private Mapper createMapperInstance(ClassLoader classLoader, String userMapper, Context context) {
+  private Mapper createMapperInstance(ClassLoader classLoader, String userMapper, Context context, String program) {
     if (context.getInputSplit() instanceof MultiInputTaggedSplit) {
       // Find the delegate Mapper from the MultiInputTaggedSplit.
       userMapper = ((MultiInputTaggedSplit) context.getInputSplit()).getMapperClassName();
     }
     try {
       return (Mapper) classLoader.loadClass(userMapper).newInstance();
     } catch (Exception e) {
+      Throwable rootCause = Throwables.getRootCause(e);
```

Comment:
@anew Like we have discussed in the past, this could indeed mean some context is missed out. But if we had to print one message in the user error from the exception being handled, the best chance of getting the real reason is to get it from the root cause. The system logs can still carry the complete exception indicating the complete context, which the user can look at. Showing the long stack trace in the User logs might not be the best experience.
I think rootCause works well in most cases. Here is one example from a use case in hand: 
2017-05-26 18:03:59,191 - ERROR [pcontroller-program:default.Usecase5_IncorrectSchema.-SNAPSHOT.workflow.DataPipelineWorkflow-88bac7f2-423d-11e7-a2f0-42010afa002e:c.c.c.i.a.r.d.AbstractProgramTwillRunnable@319] - Program runner error out.
java.lang.RuntimeException: org.apache.hadoop.mapreduce.lib.input.InvalidInputException: Input path does not exist: hdfs://10.250.0.46/users/d
	at com.google.common.base.Throwables.propagate(Throwables.java:160) ~[com.google.guava.guava-13.0.1.jar:na]
	at co.cask.cdap.internal.app.runtime.workflow.WorkflowDriver.executeAll(WorkflowDriver.java:592) ~[na:na]
	at co.cask.cdap.internal.app.runtime.workflow.WorkflowDriver.run(WorkflowDriver.java:568) ~[na:na]
	at com.google.common.util.concurrent.AbstractExecutionThreadService$1$1.run(AbstractExecutionThreadService.java:52) ~[com.google.guava.guava-13.0.1.jar:na]
	at java.lang.Thread.run(Thread.java:745) [na:1.7.0_75]
Caused by: org.apache.hadoop.mapreduce.lib.input.InvalidInputException: Input path does not exist: hdfs://10.250.0.46/users/d
	at org.apache.hadoop.mapreduce.lib.input.FileInputFormat.singleThreadedListStatus(FileInputFormat.java:323) ~[hadoop-mapreduce-client-core-2.7.1.2.3.4.7-4.jar:na]
	at org.apache.hadoop.mapreduce.lib.input.FileInputFormat.listStatus(FileInputFormat.java:265) ~[hadoop-mapreduce-client-core-2.7.1.2.3.4.7-4.jar:na]
	at org.apache.hadoop.mapreduce.lib.input.CombineFileInputFormat.getSplits(CombineFileInputFormat.java:217) ~[hadoop-mapreduce-client-core-2.7.1.2.3.4.7-4.jar:na]
	at co.cask.cdap.internal.app.runtime.batch.dataset.input.MultiInputFormat.getSplits(MultiInputFormat.java:70) ~[na:na]
	at org.apache.hadoop.mapreduce.JobSubmitter.writeNewSplits(JobSubmitter.java:301) ~[hadoop-mapreduce-client-core-2.7.1.2.3.4.7-4.jar:na]
	at org.apache.hadoop.mapreduce.JobSubmitter.writeSplits(JobSubmitter.java:318) ~[hadoop-mapreduce-client-core-2.7.1.2.3.4.7-4.jar:na]
	at org.apache.hadoop.mapreduce.JobSubmitter.submitJobInternal(JobSubmitter.java:196) ~[hadoop-mapreduce-client-core-2.7.1.2.3.4.7-4.jar:na]
	at org.apache.hadoop.mapreduce.Job$10.run(Job.java:1290) ~[hadoop-mapreduce-client-core-2.7.1.2.3.4.7-4.jar:na]
	at org.apache.hadoop.mapreduce.Job$10.run(Job.java:1287) ~[hadoop-mapreduce-client-core-2.7.1.2.3.4.7-4.jar:na]
	at java.security.AccessController.doPrivileged(Native Method) ~[na:1.7.0_75]
	at javax.security.auth.Subject.doAs(Subject.java:415) ~[na:1.7.0_75]
	at org.apache.hadoop.security.UserGroupInformation.doAs(UserGroupInformation.java:1657) ~[hadoop-common-2.7.1.2.3.4.7-4.jar:na]
	at org.apache.hadoop.mapreduce.Job.submit(Job.java:1287) ~[hadoop-mapreduce-client-core-2.7.1.2.3.4.7-4.jar:na]
	at co.cask.cdap.internal.app.runtime.batch.MapReduceRuntimeService.startUp(MapReduceRuntimeService.java:345) ~[na:na]
	at com.google.common.util.concurrent.AbstractExecutionThreadService$1$1.run(AbstractExecutionThreadService.java:47) ~[com.google.guava.guava-13.0.1.jar:na]
	at co.cask.cdap.internal.app.runtime.batch.MapReduceRuntimeService$1$1.run(MapReduceRuntimeService.java:464) ~[na:na]
	... 1 common frames omitted